File: chessmenthol/analysis/book.py
from __future__ import annotations

from pathlib import Path
from typing import Protocol

import chess
import chess.polyglot


class BookLookup(Protocol):
    def contains_move(self, board: chess.Board, move: chess.Move) -> bool: ...


class NoBook:
    """Null book: nothing is ever a book move."""

    def contains_move(self, board: chess.Board, move: chess.Move) -> bool:
        return False


class PolyglotBook:
    """Looks moves up in a Polyglot (.bin) opening book."""

    def __init__(self, path: Path):
        self._path = Path(path)

    def contains_move(self, board: chess.Board, move: chess.Move) -> bool:
        try:
            with chess.polyglot.open_reader(self._path) as reader:
                return any(entry.move == move for entry in reader.find_all(board))
        except (FileNotFoundError, IndexError, OSError):
            return False

File: frontend/src/core/book.ts
import type { Chess } from './chess';

/** Looks up whether a move is a known opening-book move. */
export interface BookLookup {
  containsMove(pos: Chess, uci: string): boolean;
}

/** Null book: nothing is ever a book move (the parity default). */
export class NoBook implements BookLookup {
  containsMove(_pos: Chess, _uci: string): boolean {
    return false;
  }
}

// PolyglotBook (reading a .bin Polyglot opening book) is deferred — not on the
// Phase 1b parity path, which uses NoBook (empty book) as in the Python default.
